### DISCUSSION TITLE
Can't resolve 'fs' - NEXT.js

### DISCUSSION THREAD

**Discussion by tonymanuel** [ben-rogerson/twin.macro#219] [Q&A] [answered]
When I import tw from "twin.macro" in my NEXT.js application, it throws me the following error. Could you please tell me how can I get rid of this error.

![Screenshot from 2020-11-29 23-03-54](https://user-images.githubusercontent.com/15573335/100549407-cd8e3980-3298-11eb-9eeb-9fd6a990e4d3.png)

**Comment by ben-rogerson** [comment, marked as answer]
Sure, check out [this next config](https://github.com/ben-rogerson/twin.examples/commit/88df17fdeb3267da5c5563352955d834d6a52008#diff-f554427811f4885204785fc4b08a6d863b9a48f78b4f3bb3a75af4f207c1089fL169) for the fix. 
New: This isn't needed in newer versions of next. If you're seeing errors, try removing your `.next` folder and restarting your dev process.

**Reply by tonymanuel** [reply]
Thanks its working now

**Reply by rodmartinezmedina** [reply]
Hello. I'm still not being able to run it. I Tried the next.config.js file from the link but still not working. Please help
I'm getting the following:

```
Unhandled Runtime Error
SyntaxError: Invalid regular expression: invalid group specifier name
Call Stack
eval

[native code]
./node_modules/twin.macro/macro.js

file:///Users/rodrigomac/WEB-DEVELOPMENT/CTAIMA/hello-ctaima/.next/static/chunks/pages/_app.js (4553:5)
__webpack_require__

/_next/static/chunks/webpack.js (873:35)
fn

file:///Users/rodrigomac/WEB-DEVELOPMENT/CTAIMA/hello-ctaima/.next/static/chunks/webpack.js (151:39)
eval

[native code]
./pages/_app.js

/_next/static/chunks/pages/_app.js (4764:5)
__webpack_require__

file:///Users/rodrigomac/WEB-DEVELOPMENT/CTAIMA/hello-ctaima/.next/static/chunks/webpack.js (873:35)
fn

/_next/static/chunks/webpack.js (151:39)
```

**Reply by ben-rogerson** [reply]
@rodmartinezmedina Do the [next examples](https://github.com/ben-rogerson/twin.examples) work for you? Perhaps you could compare against those examples to see where the issue is.

**Reply by OladetounJed** [reply]
Hello @ben-rogerson , I am getting the "can't resolve module 'fs'" error but seems the next.config.js You are referring to in the thread has been deleted.

**Reply by ben-rogerson** [reply]
Yeah that's for an older version of next - You no longer need the customisations in next.config.js.
Here are [the deleted next.config.js customisations](https://github.com/ben-rogerson/twin.examples/commit/88df17fdeb3267da5c5563352955d834d6a52008#diff-f554427811f4885204785fc4b08a6d863b9a48f78b4f3bb3a75af4f207c1089fL169).

Edit: On a recent next js project, I still needed the fix below.
Seems dependant on certain plugins/packages you use.
 
```js
// next.config.js
module.exports = {
  webpack: config => {
    // Unset client-side javascript that only works server-side
    config.resolve.fallback = { fs: false, module: false, path: false }
    return config
  },
}
```

**Reply by aroc** [reply]
Sadly, this doesn't work for me. I'm running nextjs `12.0.7` and twin `2.8.2` with webpack 5. If I add this to my webpack config, I get `can't resolve module 'os'`. If I additionally add `os: false` to the webpack config code then I get the error: `MacroError: The macro you imported from "undefined" is being executed outside the context of compilation with babel-plugin-macros. This indicates that you don't have the babel plugin "babel-plugin-macros" configured correctly. Please see the documentation for how to configure babel-plugin-macros properly: https://github.com/kentcdodds/babel-plugin-macros/blob/master/other/docs/user.md`

**Reply by aroc** [reply]
Ok using this babel config has worked for me:
https://github.com/facebook/relay/issues/2648#issuecomment-499250143

**Comment by rodmartinezmedina** [comment]
Thanks a lot for your reply. Solved it. Somehow the next.config.js was not
working at first.
Now everything's working fine.
Amazing package. Thanks so much.


Rodrigo Martinez
| Frontend Developer | UI-UX | Product Design

**Reply by ChrisClaude** [reply]
I am using Next JS 11 and webpack 5 enabled. I first got the fs error and then updated my next config with [next examples](https://github.com/ben-rogerson/twin.examples/blob/master/next-styled-components/next.config.js) but I am now getting the following error: 

`./node_modules/babel-plugin-macros/dist/index.js:9:0
Module not found: Can't resolve 'path'
null`

**Reply by agusterodin** [reply]
Did you ever find a solution?

**Reply by agusterodin** [reply]
Also that link is dead, do you remember what you changed?

**Reply by xairoo** [reply]
Same for me.

**Reply by ben-rogerson** [reply]
@Xairoo Next.js + webpack 5 (default), try this:

```js
// next.config.js
module.exports = {
  webpack: config => {
    // Unset client-side javascript that only works server-side
    config.resolve.fallback = { fs: false, module: false, path: false }
    return config
  },
}
```

**Reply by xairoo** [reply]
@ben-rogerson works, thanks! It's a bit confusing. I am using `i18next` (https://github.com/Xairoo/nextjs-i18n-static-page-starter), guess that's causing the error because of `getStaticProps` and/or `getStaticPaths` - btw. how the locales data gets loaded.

When I use `next-i18next` it just works out of the box. No need to update the `next.config.js`.

Anyway: `twin.macro` is awesome, thanks!

**Reply by Oyelowo** [reply]
I'm still getting the issue following this example: https://github.com/ben-rogerson/twin.examples/blob/master/storybook-emotion/.storybook/main.js

I'm combining twin.macro with storybook + emotions and nextjs(typescript)

Good job with this library! It is by far the most ergonomic CSS library I've worked with

**Reply by ben-rogerson** [reply]
@Oyelowo Thanks, I'll take a look into it.

**Reply by isurucuma** [reply]
![image](https://user-images.githubusercontent.com/40654309/147506151-1ee345fc-c4d7-46dd-8bb3-06df0607b829.png)
Hi I am getting this error 
I am using "fs" and "path" modules in my serverSideProps function. At first I got **"fs" is not defined** and then I tried editing config file by adding **{ fs: false, module: false, path: false }** then I got the error which is in the attachment above. Could you please help me. 

I have imported this markdown_utils.js inside my component 
![image](https://user-images.githubusercontent.com/40654309/147506659-6c2eb426-1096-4c27-81d6-c28114646056.png)

When this try to load this markdown_utils.js file from the component this error appear and component is not loaded to the browser


Thank you so much

**Reply by ben-rogerson** [reply]
Does importing path in your utils file help?

```js
const path = require('path');
```
